### PR TITLE
[Enhancement] assign morsels uniformly among ScanOperators to avoid skew

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

When morsel count is less than `dop * io_parallelism`, assign morsels to scan operators uniformly to avoid skew.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.3 branch-2.4

**Comment by kangkaisen** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by kangkaisen — DISMISSED** [review]

File: be/src/exec/scan_node.cpp [hunk 1]
```diff
@@ -63,7 +81,15 @@ StatusOr<pipeline::MorselQueueFactoryPtr> ScanNode::convert_scan_range_to_morsel
                                                     global_scan_ranges, node_id, pipeline_dop,
                                                     enable_tablet_internal_parallel, global_scan_ranges.size()));
         int scan_dop = std::min<int>(std::max<int>(1, morsel_queue->max_degree_of_parallelism()), pipeline_dop);
-        return std::make_unique<pipeline::SharedMorselQueueFactory>(std::move(morsel_queue), scan_dop);
+        int io_parallelism = scan_dop * io_tasks_per_scan_operator();
+
+        // If not so much morsels, try to assign morsel uniformly among operators to avoid data skew
+        if (scan_dop > 1 && morsel_queue->num_original_morsels() <= io_parallelism) {
+            auto morsel_queue_map = uniform_distribute_morsels(std::move(morsel_queue), scan_dop);
+            return std::make_unique<pipeline::IndividualMorselQueueFactory>(std::move(morsel_queue_map));
```
Comment (ZiheLiu) [inline be/src/exec/scan_node.cpp:89]: Only `FixedMorselQueue` need `uniform_distribute_morsels`, but `morsel_queue` here may be `PhysicalSplitMorselQueue` or `LogicalSplitMorselQueue`.

File: be/src/exec/scan_node.cpp [hunk 2]
```diff
@@ -63,7 +81,15 @@ StatusOr<pipeline::MorselQueueFactoryPtr> ScanNode::convert_scan_range_to_morsel
                                                     global_scan_ranges, node_id, pipeline_dop,
                                                     enable_tablet_internal_parallel, global_scan_ranges.size()));
         int scan_dop = std::min<int>(std::max<int>(1, morsel_queue->max_degree_of_parallelism()), pipeline_dop);
-        return std::make_unique<pipeline::SharedMorselQueueFactory>(std::move(morsel_queue), scan_dop);
+        int io_parallelism = scan_dop * io_tasks_per_scan_operator();
+
+        // If not so much morsels, try to assign morsel uniformly among operators to avoid data skew
+        if (scan_dop > 1 && morsel_queue->num_original_morsels() <= io_parallelism) {
```
Comment (silverbullet233) [inline be/src/exec/scan_node.cpp:87]: Can this be a general strategy? What would be the problem if we apply this rule when there are so many morsels?

File: be/src/exec/scan_node.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exec/scan_node.cpp]: fixed

File: be/src/exec/scan_node.cpp (hunk 2 above)
Comment (mofeiatwork) [reply be/src/exec/scan_node.cpp]: I'm afraid not. It's conflicted with tablet internal-parallel mechanism. We need to unify these strategies in the future.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by silverbullet233 — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**